### PR TITLE
Add `static_multimap::pair_contains`

### LINKED ISSUE #173: [FEA] Support `pair_contains` in `static_map` and `static_multimap`
In `static_multimap`, there are various `pair_*` APIs but not `pair_contains`. Sometimes we need it for checking the existence of key/value pairs given from input iterators.

It's great to have such API added to `static_multimap` as well as `static_map`.

### PR DESCRIPTION
**Pull request by PointKernel**
Depends on #174

Contributes to #173 

This PR adds `static_multimap::pair_contains`.

### CODE REVIEW

File: include/cuco/detail/static_multimap/device_view_impl.inl [hunk 1]
```diff
@@ -559,29 +559,33 @@ class static_multimap<Key, Value, Scope, Allocator, ProbeSequence>::device_view_
...
   __device__ __forceinline__ std::enable_if_t<uses_vector_load, bool> contains(
-    CG g, Key const& k, KeyEqual key_equal) noexcept
+    cooperative_groups::thread_block_tile<ProbeSequence::cg_size> const& g,
+    Element const& element,
+    Equal equal) noexcept
   {
-    auto current_slot = initial_slot(g, k);
+    auto current_slot = [&]() {
+      if constexpr (is_pair_contains) { return initial_slot(g, element.first); }
+      if constexpr (not is_pair_contains) { return initial_slot(g, element); }
```
Comment (ttnghia) [inline include/cuco/detail/static_multimap/device_view_impl.inl:595]:
```suggestion
      else { return initial_slot(g, element); }
```

File: include/cuco/detail/static_multimap/device_view_impl.inl [hunk 2]
```diff
@@ -591,8 +595,22 @@ class static_multimap<Key, Value, Scope, Allocator, ProbeSequence>::device_view_
...
         detail::bitwise_compare(arr[1].first, this->get_empty_key_sentinel());
-      auto const first_equals  = (not first_slot_is_empty and key_equal(arr[0].first, k));
-      auto const second_equals = (not second_slot_is_empty and key_equal(arr[1].first, k));
+      auto const first_equals = [&]() {
+        if constexpr (is_pair_contains) {
+          return not first_slot_is_empty and equal(arr[0], element);
+        }
+        if constexpr (not is_pair_contains) {
+          return not first_slot_is_empty and equal(arr[0].first, element);
+        }
```
Comment (ttnghia) [inline include/cuco/detail/static_multimap/device_view_impl.inl:612]:
```suggestion
        else {
          return not first_slot_is_empty and equal(arr[0].first, element);
        }
```

File: include/cuco/detail/static_multimap/device_view_impl.inl (hunk 1 above)
Comment (ttnghia) [reply include/cuco/detail/static_multimap/device_view_impl.inl:595]: `else` after `if constexpr` also means `else constexpr`.

File: include/cuco/detail/static_multimap/device_view_impl.inl [hunk 3]
```diff
@@ -607,29 +625,33 @@ class static_multimap<Key, Value, Scope, Allocator, ProbeSequence>::device_view_
...
   __device__ __forceinline__ std::enable_if_t<not uses_vector_load, bool> contains(
-    CG g, Key const& k, KeyEqual key_equal) noexcept
+    cooperative_groups::thread_block_tile<ProbeSequence::cg_size> const& g,
+    Element const& element,
+    Equal equal) noexcept
   {
-    auto current_slot = initial_slot(g, k);
+    auto current_slot = [&]() {
+      if constexpr (is_pair_contains) { return initial_slot(g, element.first); }
+      if constexpr (not is_pair_contains) { return initial_slot(g, element); }
```
Comment (ttnghia) [inline include/cuco/detail/static_multimap/device_view_impl.inl:661]: Similar to above.

Comment (ttnghia) [reply include/cuco/detail/static_multimap/device_view_impl.inl:661]: And also all other places.

File: include/cuco/detail/static_multimap/static_multimap.inl [hunk 4]
```diff
@@ -543,7 +568,24 @@ static_multimap<Key, Value, Scope, Allocator, ProbeSequence>::device_view::conta
...
+template <typename Key,
+          typename Value,
+          cuda::thread_scope Scope,
+          typename Allocator,
+          class ProbeSequence>
+template <typename PairEqual>
+__device__ __forceinline__ bool
+static_multimap<Key, Value, Scope, Allocator, ProbeSequence>::device_view::pair_contains(
+  cooperative_groups::thread_block_tile<ProbeSequence::cg_size> const& g,
+  value_type const& p,
```
Comment (ttnghia) [inline include/cuco/detail/static_multimap/static_multimap.inl]: `value_type` is wrong. Should be `Element const& p`.

File: include/cuco/static_multimap.cuh [hunk 5]
```diff
@@ -834,6 +860,28 @@ class static_multimap {
...
+     * @param g The Cooperative Group used to perform the contains operation
+     * @param k The pair to search for
+     * @param pair_equal The binary callable used to compare input pair and slot content
+     * for equality
+     * @return A boolean indicating whether the input pair was inserted in the map
+     */
+    template <typename PairEqual = thrust::equal_to<value_type>>
+    __device__ __forceinline__ bool pair_contains(
+      cooperative_groups::thread_block_tile<ProbeSequence::cg_size> const& g,
+      value_type const& p,
```
Comment (ttnghia) [inline include/cuco/static_multimap.cuh:882]: `Element const& element,`

File: include/cuco/detail/static_multimap/static_multimap.inl [hunk 6]
```diff
@@ -543,7 +568,24 @@ static_multimap<Key, Value, Scope, Allocator, ProbeSequence>::device_view::conta
...
+  constexpr bool is_pair_contains = false;
+  return impl_.contains<is_pair_contains, uses_vector_load()>(g, k, key_equal);
+}
+
+template <typename Key,
+          typename Value,
+          cuda::thread_scope Scope,
+          typename Allocator,
+          class ProbeSequence>
+template <typename PairEqual>
```
Comment (ttnghia) [inline include/cuco/detail/static_multimap/static_multimap.inl]:
```suggestion
template <typename Element, typename PairEqual>
```

File: include/cuco/static_multimap.cuh [hunk 7]
```diff
@@ -834,6 +860,28 @@ class static_multimap {
...
+     * `contains` at moderate to high load factors.
+     *
+     * @tparam PairEqual Binary callable type
+     * @param g The Cooperative Group used to perform the contains operation
+     * @param k The pair to search for
+     * @param pair_equal The binary callable used to compare input pair and slot content
+     * for equality
+     * @return A boolean indicating whether the input pair was inserted in the map
+     */
+    template <typename PairEqual = thrust::equal_to<value_type>>
```
Comment (ttnghia) [inline include/cuco/static_multimap.cuh]:
```suggestion
    template <typename Element, typename PairEqual = thrust::equal_to<value_type>>
```

File: include/cuco/static_multimap.cuh (hunk 5 above)
Comment (ttnghia) [inline include/cuco/static_multimap.cuh]:
```suggestion
      Element const& element,
```

**Comment by ttnghia** [comment]
I tested locally with my PR. Will work with this fix: https://github.com/NVIDIA/cuCollections/commit/331b24837ae515ac0ce92fac4a01053898b04b47.
This also needs more cleanup and fixing for docs.

File: include/cuco/static_multimap.cuh [hunk 8]
```diff
@@ -834,6 +860,28 @@ class static_multimap {
...
+    __device__ __forceinline__ bool pair_contains(
+      cooperative_groups::thread_block_tile<ProbeSequence::cg_size> const& g,
+      value_type const& p,
+      PairEqual pair_equal = PairEqual{}) noexcept;
```
Comment (ttnghia) [inline include/cuco/static_multimap.cuh]: In addition, we also need to add the version of `pair_contains` here without the `thread_block_tile` parameter.

Comment (PointKernel) [reply include/cuco/static_multimap.cuh]: Why is that?

Comment (ttnghia) [reply include/cuco/static_multimap.cuh]: This is exposed to the users, right? So the users don't have that parameter.

Comment (PointKernel) [reply include/cuco/static_multimap.cuh]: Unlike single map, multimap needs to handle duplicates. That's the reason that we provide CG APIs only in multimap. Non-CG APIs are provided in single map but are also rarely used. Can you provide more details on your use case i.e. why would non-CG implementation is preferred over the CG ones?

Comment (PointKernel) [reply include/cuco/static_multimap.cuh]: >  So the users don't have that parameter.

users can create a CG e.g. https://godbolt.org/z/o6jYYv7Ys

Comment (ttnghia) [inline include/cuco/static_multimap.cuh]:
```suggestion
      PairEqual pair_equal = PairEqual{}) const noexcept;
```

Comment (ttnghia) [reply include/cuco/static_multimap.cuh]: I believe that the line 861 above also needs to be fixed like this. Otherwise, calling the function on const object will cause compile error.

File: include/cuco/static_multimap.cuh [hunk 9]
```diff
@@ -280,15 +280,41 @@ class static_multimap {
...
+
+  /**
+   * @brief Indicates whether the pairs in the range `[first, last)` are contained in the map.
+   *
+   * Stores `true` or `false` to `(output + i)` indicating if the pair `*(first + i)` exists in
+   * the map.
+   *
+   * @tparam InputIt Device accessible random access input iterator where
+   * `std::is_convertible<std::iterator_traits<InputIt>::value_type,
+   * static_multimap<K, V>::value_type>` is `true`
```
Comment (jrhemstad) [inline include/cuco/static_multimap.cuh]: The convertibility isn't required with heterogeneous lookup.

File: include/cuco/static_multimap.cuh [hunk 10]
```diff
@@ -280,15 +280,41 @@ class static_multimap {
...
+   * @tparam OutputIt Device accessible output iterator whose `value_type` is convertible from
+   * `bool`
```
Comment (jrhemstad) [inline include/cuco/static_multimap.cuh]:
```suggestion
   * @tparam OutputIt Device accessible output iterator assignable from `bool`
```

File: include/cuco/detail/static_multimap/static_multimap.inl (hunk 6 above)
Comment (jrhemstad) [reply include/cuco/detail/static_multimap/static_multimap.inl]: `ProbePair` instead of `Element`.

File: include/cuco/detail/probe_sequence_impl.cuh [hunk 11]
```diff
@@ -195,8 +197,8 @@ class linear_probing_impl
    * @return Pointer to the initial slot for `k`
    */
   template <typename ProbeKey>
-  __device__ __forceinline__ iterator
-  initial_slot(cooperative_groups::thread_block_tile<cg_size> const& g, ProbeKey const& k) noexcept
+  __device__ __forceinline__ iterator initial_slot(
```
Comment (jrhemstad) [inline include/cuco/detail/probe_sequence_impl.cuh]: The `const` overload should return `const_iterator`.

File: include/cuco/static_multimap.cuh [hunk 12]
```diff
@@ -293,6 +292,35 @@ class static_multimap {
...
+   * @tparam OutputIt Device accessible output iterator assignable from `bool`
+   * @tparam PairEqual Binary callable type used to compare input pair and slot content for equality
+   *
+   * @param first Beginning of the sequence of pairs
+   * @param last End of the sequence of pairs
+   * @param output_begin Beginning of the output sequence indicating whether each pair is present
+   * @param pair_equal The binary function to compare input pair and slot content for equality
+   * @param stream CUDA stream used for contains
+   */
+  template <typename InputIt, typename OutputIt, typename PairEqual = thrust::equal_to<value_type>>
```
Comment (jrhemstad) [inline include/cuco/static_multimap.cuh]: `pair_*` functions should never default the equality comparator as there is no point in using such APIs if one does not specify a custom equality function.

File: tests/static_multimap/pair_function_test.cu [hunk 13]
```diff
@@ -56,6 +57,21 @@ __inline__ void test_pair_functions(Map& map, PairIt pair_begin, std::size_t num
                       return cuco::pair_type<Key, Value>{i, i};
                     });
 
+  SECTION("Tests of pair_contains.")
```
Comment (jrhemstad) [inline tests/static_multimap/pair_function_test.cu]: This section description should be improved.

File: include/cuco/detail/static_multimap/device_view_impl.inl (hunk 3 above)
Comment (PointKernel) [reply include/cuco/detail/static_multimap/device_view_impl.inl:661]: Now I recall, we need this to silence the [compiler warnings with cuda-11.0/11.2](https://gpuci.gpuopenanalytics.com/blue/organizations/jenkins/nvidia%2Fgpuci%2FcuCollections%2Fprb%2FcuCollections-gpu-build/detail/cuCollections-gpu-build/558/pipeline). Reverting back to the `if constexpr (not ...)` expression.

**Comment by ttnghia** [comment]
> Closes #173

Hey, since #173 now saying "Support `pair_contains` in `static_map` and `static_multimap`", this should not be enough to close that.

**Comment by PointKernel** [comment]
> > Closes #173
> 
> Hey, since #173 now saying "Support `pair_contains` in `static_map` and `static_multimap`", this should not be enough to close that.

I'm aware of that. Don't worry. :)

**Comment by ttnghia** [comment]
Confirm that the latest commit works with my code.

File: tests/static_multimap/pair_function_test.cu [hunk 14]
```diff
@@ -124,11 +140,11 @@ TEMPLATE_TEST_CASE_SIG(
                           cuco::linear_probing<1, cuco::detail::MurmurHash3_32<Key>>>
       map{
         num_pairs * 2, cuco::sentinel::empty_key<Key>{-1}, cuco::sentinel::empty_value<Value>{-1}};
-    test_pair_functions<Key, Value>(map, d_pairs.begin(), num_pairs);
+    test_pair_functions<Key, Value>(map, d_pairs.data().get(), num_pairs);
```
Comment (jrhemstad) [inline tests/static_multimap/pair_function_test.cu]: Why was this change needed?

Comment (PointKernel) [reply tests/static_multimap/pair_function_test.cu]: This is to get rid of the dereference error at https://github.com/PointKernel/cuCollections/blob/add-multimap-pair-contains/include/cuco/detail/static_multimap/device_view_impl.inl#L660
```bash
/home/yunsongw/Work/cuCollections/include/cuco/detail/static_multimap/device_view_impl.inl(660): error: class "thrust::device_reference<cuco::pair<int64_t, int64_t>>" has no member "first"
          detected during:
            instantiation of "std::enable_if_t<<expression>, __nv_bool> cuco::static_multimap<Key, Value, Scope, Allocator, ProbeSequence>::device_view_impl::contains<is_pair_contains,uses_vector_load,ProbeT,Equal>(const cooperative_groups::__v1::thread_block_tile<ProbeSequence::cg_size, void> &, const ProbeT &, Equal) const noexcept [with Key=int64_t, Value=int64_t, Scope=cuda::std::__4::__detail::thread_scope_device, Allocator=cuco::cuda_allocator<char>, ProbeSequence=cuco::double_hashing<8U, cuco::detail::MurmurHash3_32<int64_t>, cuco::detail::MurmurHash3_32<int64_t>>, is_pair_contains=true, uses_vector_load=false, ProbeT=thrust::device_reference<cuco::pair<int64_t, int64_t>>, Equal=pair_equal<int64_t, int64_t>]" 
```

Comment (jrhemstad) [reply tests/static_multimap/pair_function_test.cu]: I suspected it was something like that. 

cuco functions need to be able to work on Thrust iterators directly and we can't assume inputs will always be raw pointers like `data().get()`. 

I'll need to look further into the code here, but I suspect we need to do some coaxing of the type into the shape we expect.

Comment (PointKernel) [reply tests/static_multimap/pair_function_test.cu]: The iterator is dereferenced [here](https://github.com/PointKernel/cuCollections/blob/add-multimap-pair-contains/include/cuco/detail/static_multimap/kernels.cuh#L186) by using `*`.

Comment (jrhemstad) [reply tests/static_multimap/pair_function_test.cu]: As we discussed offline, a short term fix is we can replace that line with:
```
std::iterator_traits<InputIt>::value_type element = *(first + idx);
```

Comment (PointKernel) [reply tests/static_multimap/pair_function_test.cu]: Fixed.

**Review by jrhemstad — APPROVED** [review]
